services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java

```diff
@@ -110,7 +111,20 @@ public boolean isValid(RealmModel realm, UserModel user, CredentialInput credent
             return true;
         } else if (OTPCredentialModel.TOTP.equals(credentialData.getSubType())) {
             TimeBasedOTP validator = new TimeBasedOTP(credentialData.getAlgorithm(), credentialData.getDigits(), credentialData.getPeriod(), policy.getLookAheadWindow());
-            return validator.validateTOTP(challengeResponse, secretData.getValue().getBytes(StandardCharsets.UTF_8));
+
+            boolean validReusedOtp = policy.isCodeReusable() || !challengeResponse.equals(secretData.getLastValue());
+            if (!validReusedOtp) {
+                logger.debug("It is forbidden to use the same OTP code twice. Please wait for the next one.");
```

Comment:
I would remove this log entry as it is not shown to the user. If we want to enable logging, we should also be adding to other places too like when the OTP is just invalid.

Reply:
Ok, agree.